### PR TITLE
Add Coherence based caching implementation for Micronaut

### PR DESCRIPTION
**Pull request by vasac**

### CODE REVIEW

**Review by aseovic — CHANGES_REQUESTED** [review]
Looks good, but there are some minor changes to be made in async cache implementation, to offload response processing from a service thread to an executor service thread.

We also need to discuss and decide if the caches should be user-managed, as is the case now, or system-managed and defined in our system cache config.

File: coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceCacheManager.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+                                 Coherence coherence) {
+        this.conversionService = conversionService;
+        this.executorService = executorService;
+        this.session = coherence.getSession();
+    }
+
+    @NonNull
+    @Override
+    public SyncCache<NamedCache<Object, Object>> getCache(String name) {
+        NamedCache<Object, Object> nativeCache = session.getCache(name);
```
[the remark this answers is not in the document]
Comment (aseovic) [reply coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceCacheManager.java:58]: We should allow users to configure a `coherence.caching.prefix` prefix that will be used for all cache names, and will be blank by default.

File: coherence-cache/build.gradle [hunk 2]
```diff
@@ -0,0 +1,19 @@
+dependencies {
+    annotationProcessor "io.micronaut:micronaut-inject-java"
+    annotationProcessor "io.micronaut:micronaut-inject-groovy"
+    annotationProcessor "io.micronaut.docs:micronaut-docs-asciidoc-config-props:$micronautDocsVersion"
+
+    implementation project(":coherence")
+    api "io.micronaut:micronaut-inject"
+    api "io.micronaut.cache:micronaut-cache-core"
+    api "com.oracle.coherence.ce:coherence"
```
[the remark this answers is not in the document]
Comment (alvarosanchez) [reply coherence-cache/build.gradle:9]: Is the idea of `compileOnly` that users can choose between community and enterprise editions? If so, when this is integrated in Micronaut Launch, you would need to create 2 different features that add the correct dependency in each case

[the remark this answers is not in the document]
Comment (vasac) [reply coherence-cache/build.gradle:9]: Yeah, we'll add Coherence dependency too in Micronaut Launch.

File: coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceCacheManager.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (vasac) [reply coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceCacheManager.java:58]: Left this as it is as `micronaut-cache` test checks whether provided cache name matches requested cache name.

File: coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceAsyncCache.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Supplier;
+
+import javax.annotation.Nonnull;
+
+/**
+ * A {@link io.micronaut.cache.SyncCache} implementation based on Coherence.
```
[the remark this answers is not in the document]
Comment (vasac) [reply coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceAsyncCache.java:37]: Fixed.

File: coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceAsyncCache.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+public class CoherenceAsyncCache implements AsyncCache<NamedCache<Object, Object>> {
+
+    private final ConversionService<?> conversionService;
+    private final AsyncNamedCache<Object, Object> asyncCache;
+    private final ExecutorService executorService;
+
+    /**
+     * @param conversionService the conversion service
+     * @param nativeCache       the native cache
+     * @param executorService   managers the pool of executors
```
[the remark this answers is not in the document]
Comment (vasac) [reply coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceAsyncCache.java:51]: Fixed.

File: coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceAsyncCache.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+                future.completeExceptionally(throwable);
+            }
+        });
+        return future;
+    }
+
+    @Override
+    public <T> CompletableFuture<T> get(@NonNull Object key, @NonNull Argument<T> requiredType, @NonNull Supplier<T> supplier) {
+        ArgumentUtils.requireNonNull("key", key);
+        return get(key, requiredType).thenApply(existingValue -> {
```
[the remark this answers is not in the document]
Comment (vasac) [reply coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceAsyncCache.java:83]: Fixed.

File: coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceAsyncCache.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        this.asyncCache = nativeCache.async();
+        this.executorService = executorService;
+    }
+
+    @Nonnull
+    @Override
+    public <T> CompletableFuture<Optional<T>> get(@NonNull Object key, @NonNull Argument<T> requiredType) {
+        ArgumentUtils.requireNonNull("key", key);
+        CompletableFuture<Optional<T>> future = new CompletableFuture<>();
+        asyncCache.get(key).whenComplete((response, throwable) -> {
```
[the remark this answers is not in the document]
Comment (vasac) [reply coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceAsyncCache.java:66]: Fixed.

File: coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceAsyncCache.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+       });
+    }
+
+    @SuppressWarnings("unchecked")
+    @NonNull
+    @Override
+    public <T> CompletableFuture<Optional<T>> putIfAbsent(@NonNull Object key, @NonNull T value) {
+        ArgumentUtils.requireNonNull("key", key);
+        ArgumentUtils.requireNonNull("value", value);
+        return asyncCache.putIfAbsent(key, value).thenApply(existingValue -> {
```
[the remark this answers is not in the document]
Comment (vasac) [reply coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceAsyncCache.java:100]: Fixed.

File: coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceAsyncCache.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+            return conversionService.convert(existingValue, aClass);
+        });
+    }
+
+    @Override
+    public CompletableFuture<Boolean> put(@NonNull Object key, @NonNull Object value) {
+        ArgumentUtils.requireNonNull("key", key);
+        ArgumentUtils.requireNonNull("value", value);
+        CompletableFuture<Boolean> future = new CompletableFuture<>();
+        asyncCache.put(key, value).whenComplete((response, throwable) -> {
```
[the remark this answers is not in the document]
Comment (vasac) [reply coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceAsyncCache.java:114]: Fixed.

File: coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceAsyncCache.java [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+                future.completeExceptionally(throwable);
+            }
+        });
+        return future;
+    }
+
+    @Override
+    public CompletableFuture<Boolean> invalidateAll() {
+        CompletableFuture<Boolean> future = new CompletableFuture<>();
+        asyncCache.clear().thenRun(() -> future.complete(true));
```
[the remark this answers is not in the document]
Comment (vasac) [reply coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceAsyncCache.java:141]: Fixed.

File: coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceSyncCache.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+import io.micronaut.cache.AbstractMapBasedSyncCache;
+import io.micronaut.cache.AsyncCache;
+import io.micronaut.core.convert.ConversionService;
+
+import java.util.concurrent.ExecutorService;
+
+/**
+ * A {@link AsyncCache} implementation based on Coherence.
```
[the remark this answers is not in the document]
Comment (vasac) [reply coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceSyncCache.java:29]: Fixed.

File: coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceAsyncCache.java [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+            }
+        });
+        return future;
+    }
+
+    @Override
+    public CompletableFuture<Boolean> invalidate(Object key) {
+        ArgumentUtils.requireNonNull("key", key);
+        CompletableFuture<Boolean> future = new CompletableFuture<>();
+        asyncCache.remove(key).whenComplete((response, throwable) -> {
```
[the remark this answers is not in the document]
Comment (vasac) [reply coherence-cache/src/main/java/io/micronaut/cache/coherence/CoherenceAsyncCache.java]: Fixed.

**Review by aseovic — APPROVED** [review]
LGTM

**Pull request merged**